### PR TITLE
🐛 Generate etcd client if at least one member is healthy

### LINKED ISSUE #3839: etcdClientGenerator.forNodes fails if one etcd member is down
**What steps did you take and what happened:**
I created a 3CP control plane with KCP
Took down one etcd member by deleting etcd static Pod
Run invoking `etcdClientGenerator.forNodes` and got the following error

```
failed to get etcdStatus for workload cluster test1: failed to create
      etcd client: failed to dial endpoint etcd-test1-control-plane-5mfct with maintenance
      client: context deadline exceeded
```

**What did you expect to happen:**
I was expecting `etcdClientGenerator.forNodes` being able to connect to etcd given that two etcd members are still up.

**Anything else you would like to add:**
`etcdClientGenerator.forLeader` works

/kind bug

### PR DESCRIPTION
**Pull request by shysank**
**What this PR does / why we need it**:
This pr fixes the bug where `etcdClientGenerator.forNodes` failed even when some(not all) etcd members were up. It also does some refactoring for writing unit tests. There are 2 commits:
1. 3417409 - Refacoring and unit tests
2. fbfdf6e2ce8be3f31ff20330508af3e707c4b7c9 - `etcdClientGenerator.forNodes` fix

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3839

### CODE REVIEW

**Comment by vincepri** [comment]
@shysank Please fix the PR title

File: controlplane/kubeadm/internal/etcd_client_generator.go [hunk 1]
```diff
@@ -34,22 +34,41 @@ import (
...
+func (c *etcdClientGenerator) forNodes(ctx context.Context, nodes []corev1.Node) (*etcd.Client, error) {
+	var errs []error
+	for _, node := range nodes {
+		endpoints := []string{staticPodName("etcd", node.Name)}
+		client, err := c.getClientFactory()(ctx, endpoints)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		return client, err
```
Comment (vincepri) [inline controlplane/kubeadm/internal/etcd_client_generator.go:69]:
```suggestion
		return client, nil
```

File: controlplane/kubeadm/internal/etcd_client_generator.go [hunk 2]
```diff
@@ -34,22 +34,41 @@ import (
...
+}
 
-	p := proxy.Proxy{
-		Kind:       "pods",
-		Namespace:  metav1.NamespaceSystem,
-		KubeConfig: c.restConfig,
-		TLSConfig:  c.tlsConfig,
-		Port:       2379,
+// forNodes takes a list of nodes and returns a client for the first one that connects
+func (c *etcdClientGenerator) forNodes(ctx context.Context, nodes []corev1.Node) (*etcd.Client, error) {
```
Comment (vincepri) [inline controlplane/kubeadm/internal/etcd_client_generator.go:60]: Sidenote, this method name is super confusing and I had to read the comment above to understand what it does. I originally assumed that `forNodes` would return a client connecting (pooling) to all of the pods.

How about renaming it to something like `forFirstAvailable`? Or something similar, not sure which one is better at this points 😄 

cc @sedefsavas @randomvariable

Comment (shysank) [reply controlplane/kubeadm/internal/etcd_client_generator.go:60]: +1 for renaming

**Comment by fabriziopandini** [comment]
lgtm pending @vincepri comments

File: controlplane/kubeadm/internal/etcd_client_generator.go (hunk 2 above)
Comment (randomvariable) [reply controlplane/kubeadm/internal/etcd_client_generator.go:60]: Ditto

**Review by vincepri — APPROVED** [review]
/lgtm
/assign @randomvariable @CecileRobertMichon @ncdc

**Comment by ncdc** [comment]
/hold

I have at least 1 change I'd like to request... will get that review in soon.

File: controlplane/kubeadm/internal/etcd_client_generator.go [hunk 3]
```diff
@@ -32,30 +32,49 @@ import (
...
+	return c.clientFactory
+}
 
-	p := proxy.Proxy{
-		Kind:       "pods",
-		Namespace:  metav1.NamespaceSystem,
-		KubeConfig: c.restConfig,
-		TLSConfig:  c.tlsConfig,
-		Port:       2379,
+// forNodes takes a list of nodes and returns a client for the first one that connects
```
Comment (ncdc) [inline controlplane/kubeadm/internal/etcd_client_generator.go:57]:
```suggestion
// forFirstAvailableNode takes a list of nodes and returns a client for the first one that connects.
```

**Review by ncdc — COMMENTED** [review]
I'm not a huge fan of this syntax

```go
c.getClientFactory()(ctx, endpoints)
```

where you invoke a function that returns a function, which you then immediately invoke... In other projects, I've used a pattern like this:

```go
type widgetCreator func() *Widget

type Foo struct {
  ...
  createWidget widgetCreator
}

func NewFoo() *Foo {
  f := &Foo{
    ...
  }

  // do some setup if needed

  wc := func() *Widget {
    return ...
  }

  f.createWidget = wc
}

func (f *Foo) DoSomething() {
  widget := f.createWidget()
}
```

and then expect consumers will call `NewFoo()` to get a foo. Because the function pointer field is phrased as `<verb><noun>` e.g. `createWidget`, I think it's easier to parse `f.createWidget()` vs. something like `f.getWidgetFactory()()`.

For unit tests, you can have them call `NewFoo` and then immediately replace `createWidget` with a fake/test implementation.

Food for thought - you don't have to change things, but this is why I put the hold on it.

Please do fix the godoc though 😄

**Comment by randomvariable** [comment]
lgtm from my end

File: controlplane/kubeadm/internal/etcd_client_generator.go [hunk 4]
```diff
@@ -30,32 +30,52 @@ import (
...
+		endpoints := []string{staticPodName("etcd", name)}
+		client, err := c.createClient(ctx, endpoints)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		return client, nil
 	}
-	return etcd.NewClient(ctx, endpoints, p, c.tlsConfig)
+	return nil, errors.Wrap(kerrors.NewAggregate(errs), "could not establish a connection to any etcd node")
```
Comment (fabriziopandini) [inline controlplane/kubeadm/internal/etcd_client_generator.go:70]: nit, add the list of nodeNames to the error message

**Review by fabriziopandini — COMMENTED** [review]
/lgtm 
(a small nit, not blocking)

File: controlplane/kubeadm/internal/etcd_client_generator_test.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+	"crypto/tls"
+	"errors"
+	"strings"
+	"testing"
+
+	. "github.com/onsi/gomega"
+
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/etcdserver/etcdserverpb"
+
```
Comment (vincepri) [inline controlplane/kubeadm/internal/etcd_client_generator_test.go:30]:
```suggestion
```

**Review by vincepri — APPROVED** [review]
/approve

File: controlplane/kubeadm/internal/etcd_client_generator.go (hunk 4 above)
Comment (ncdc) [reply controlplane/kubeadm/internal/etcd_client_generator.go:70]: @shysank do you think you can add this node names soon?

Comment (ncdc) [reply controlplane/kubeadm/internal/etcd_client_generator.go:70]: Actually let's get this in & we can adjust the error message later if needed, considering this has been on hold a while & Fabrizio's request is non-blocking.

**Pull request merged**